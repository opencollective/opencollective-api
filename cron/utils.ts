import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
