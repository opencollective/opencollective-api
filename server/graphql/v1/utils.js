import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
