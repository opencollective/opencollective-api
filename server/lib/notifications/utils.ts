import sanitizeHtml from 'sanitize-html';

import { sanitizerOptions as updateSanitizerOptions } from '../../models/Update';

const constructPreviewImageURL = (service: string, id: string) => {
  if (service === 'youtube' && id.match('[a-zA-Z0-9_-]{11}')) {
    return `https://img.youtube.com/vi/${id}/0.jpg`;
  } else if (service === 'anchorFm') {
    return `https://opencollective.com/static/images/anchor-fm-logo.png`;
  } else {
    return null;
  }
};

const parseServiceLink = (videoLink: string) => {
  const regexps = {
    youtube: new RegExp(
      '(?:https?://)?(?:www\\.)?youtu(?:\\.be/|be(-nocookie)?\\.com/\\S*(?:watch|embed)(?:(?:(?=/[^&\\s?]+(?!\\S))/)|(?:\\S*v=|v/)))([^&\\s?]+)',
      'i',
    ),
    anchorFm: /^(http|https)?:\/\/(www\.)?anchor\.fm\/([^/]+)(\/embed)?(\/episodes\/)?([^/]+)?\/?$/,
  };
  for (const service in regexps) {
    videoLink = videoLink.replace('/?showinfo=0', '');
    const matches = regexps[service].exec(videoLink);
    if (matches) {
      if (service === 'anchorFm') {
        const podcastName = matches[3];
        const episodeId = matches[6];
        const podcastUrl = `${podcastName}/embed`;
        return { service, id: episodeId ? `${podcastUrl}/episodes/${episodeId}` : podcastUrl };
      } else {
        return { service, id: matches[matches.length - 1] };
      }
    }
  }
  return {};
};

export const replaceVideosByImagePreviews = (html: string) => {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    transformTags: {
      ...updateSanitizerOptions.transformTags,
      iframe: (tagName, attribs) => {
        if (!attribs.src) {
          return '';
        }
        const { service, id } = parseServiceLink(attribs.src);
        const imgSrc = constructPreviewImageURL(service, id);
        if (imgSrc) {
          return {
            tagName: 'img',
            attribs: {
              src: imgSrc,
              alt: `${service} content`,
            },
          };
        } else {
          return '';
        }
      },
    },
  };
  return sanitizeHtml(html, sanitizerOptions);
};
