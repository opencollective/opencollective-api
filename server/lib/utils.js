import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { URL } from 'url';

import Promise from 'bluebird';
import config from 'config';
import pdf from 'html-pdf';
import { filter, get, isEqual, padStart, sumBy } from 'lodash';

import { ZERO_DECIMAL_CURRENCIES } from '../constants/currencies';

import errors from './errors';
import handlebars from './handlebars';

const { BadRequest } = errors;

export function addParamsToUrl(url, obj) {
  const u = new URL(url);
  Object.keys(obj).forEach(key => {
    u.searchParams.set(key, obj[key]);
  });
  return u.href;
}

// source: https://stackoverflow.com/questions/8498592/extract-hostname-name-from-string
function extractHostname(url) {
  let hostname;
  // find & remove protocol (http, ftp, etc.) and get hostname

  if (url.indexOf('://') > -1) {
    hostname = url.split('/')[2];
  } else {
    hostname = url.split('/')[0];
  }

  // find & remove port number
  hostname = hostname.split(':')[0];
  // find & remove "?"
  hostname = hostname.split('?')[0];

  return hostname;
}

export function getDomain(url = '') {
  let domain = extractHostname(url);
  const splitArr = domain.split('.'),
    arrLen = splitArr.length;

  // extracting the root domain here
  // if there is a subdomain
  if (arrLen > 2) {
    domain = `${splitArr[arrLen - 2]}.${splitArr[arrLen - 1]}`;
    // check to see if it's using a Country Code Top Level Domain (ccTLD) (i.e. ".me.uk")
    if (splitArr[arrLen - 1].length === 2 && splitArr[arrLen - 1].length === 2) {
      // this is using a ccTLD
      domain = `${splitArr[arrLen - 3]}.${domain}`;
    }
  }
  return domain;
}

/**
 * Gives the number of days between two dates
 */
export const days = (d1, d2 = new Date()) => {
  const oneDay = 24 * 60 * 60 * 1000; // hours*minutes*seconds*milliseconds
  return Math.round(Math.abs((d1.getTime() - d2.getTime()) / oneDay));
};

export const flattenArray = arr => {
  return arr.reduce((flat, toFlatten) => {
    return flat.concat(Array.isArray(toFlatten) ? flattenArray(toFlatten) : toFlatten);
  }, []);
};

/**
 * Returns stats for each tier compared to previousMonth
 *
 * @PRE:
 *  - tiers: array of Tier: [ { name, interval, users: [ { id, totalDonations, firstDonation, lastDonation } ] } ]
 *  - startDate, endDate: boundaries for lastMonth
 *
 * @POST: { stats, tiers }
 *  - stats.backers.lastMonth: number of backers who were active by endDate
 *  - stats.backers.previousMonth: number of backers who were active by startDate
 *  - stats.backers.new: the number of backers whose first donation was after startDate
 *  - stats.backers.lost: the number of backers who were active before startDate, but stopped being active
 *  - tiers: tiers with users sorted by totalDonations
 */
export const getTiersStats = (tiers, startDate, endDate) => {
  const backersIds = {};
  const stats = { backers: {} };

  const rank = user => {
    if (user.isNew) {
      return 1;
    }
    if (user.isLost) {
      return 2;
    }
    return 3;
  };

  stats.backers.lastMonth = 0;
  stats.backers.previousMonth = 0;
  stats.backers.new = 0;
  stats.backers.lost = 0;

  // We only keep the tiers that have at least one user
  tiers = tiers.filter(tier => {
    if (get(tier, 'dataValues.users') && get(tier, 'dataValues.users').length > 0) {
      return true;
    } else {
      return false;
    }
  });

  // We sort tiers by number of users ASC
  tiers.sort((a, b) => b.amount - a.amount);

  return Promise.map(tiers, tier => {
    const backers = get(tier, 'dataValues.users');
    let index = 0;

    // We sort backers by total donations DESC
    backers.sort((a, b) => b.totalDonations - a.totalDonations);

    return Promise.filter(backers, backer => {
      if (backersIds[backer.id]) {
        return false;
      }
      backersIds[backer.id] = true;

      backer.index = index++;
      return Promise.all([tier.isBackerActive(backer, endDate), tier.isBackerActive(backer, startDate)]).then(
        results => {
          backer.activeLastMonth = results[0];
          backer.activePreviousMonth = backer.firstDonation < startDate && results[1];
          if (tier.name.match(/sponsor/i)) {
            backer.isSponsor = true;
          }
          if (backer.firstDonation > startDate) {
            backer.isNew = true;
            stats.backers.new++;
          }
          if (backer.activePreviousMonth && !backer.activeLastMonth) {
            backer.isLost = true;
            stats.backers.lost++;
          }
          if (backer.activePreviousMonth) {
            stats.backers.previousMonth++;
          }
          if (backer.activeLastMonth) {
            stats.backers.lastMonth++;
            return true;
          } else if (backer.isLost) {
            return true;
          }
        },
      );
    }).then(backers => {
      backers.sort((a, b) => {
        if (rank(a) > rank(b)) {
          return 1;
        }
        if (rank(a) < rank(b)) {
          return -1;
        }
        return a.index - b.index; // make sure we keep the original order within a tier (typically totalDonations DESC)
      });

      tier.activeBackers = backers.filter(b => !b.isLost);

      return tier;
    });
  }).then(tiers => {
    return { stats, tiers };
  });
};

/**
 * export data to CSV
 * @param {*} data
 * @param {*} attributes
 * @param {*} getColumnName
 * @param {*} processValue
 */
export function exportToCSV(data, attributes, getColumnName = attr => attr, processValue = (attr, val) => val) {
  const lines = [];

  lines.push(`"${attributes.map(getColumnName).join('","')}"`); // Header

  const getLine = row => {
    const cols = [];
    attributes.map(attr => {
      cols.push(`${processValue(attr, get(row, attr) || '')}`.replace(/\"/g, '"'));
    });
    return `"${cols.join('","')}"`;
  };

  data.map(row => {
    lines.push(getLine(row));
  });
  return lines.join('\n');
}

/**
 * export transactions to PDF
 */
export function exportToPDF(template, data, options) {
  options = options || {};
  options.paper = options.paper || 'Letter'; // Letter for US or A4 for Europe

  let paperSize;

  switch (options.paper) {
    case 'A4':
      paperSize = {
        width: '210mm',
        height: '297mm',
        margin: {
          top: '10mm',
          left: '10mm',
        },
      };
      break;
    case 'Letter':
    default:
      paperSize = {
        width: '8.5in',
        height: '11in',
        margin: {
          top: '0.4in',
          left: '0.4in',
        },
      };
      break;
  }

  data.paperSize = paperSize;
  options.paperSize = paperSize;

  const templateFilepath = path.resolve(__dirname, `../../templates/pdf/${template}.hbs`);
  const source = fs.readFileSync(templateFilepath, 'utf8');
  const render = handlebars.compile(source);

  const html = render(data);

  if (options.format === 'html') {
    return Promise.resolve(html);
  }
  options.format = options.paper;

  options.timeout = 60000;

  return new Promise((resolve, reject) => {
    pdf.create(html, options).toBuffer((err, buffer) => {
      if (err) {
        return reject(err);
      }
      return resolve(buffer);
    });
  });
}

/**
 * Default host id, set this for new collectives created through our flow
 *
 * @param {"opensource" | null} category of the collective
 */
export const defaultHostCollective = category => {
  if (config.env === 'production' || config.env === 'staging') {
    if (category === 'opensource') {
      return { id: 772, CollectiveId: 11004, ParentCollectiveId: 83 }; // Open Source Host Collective
    } else if (category === 'foundation') {
      return { CollectiveId: 11049 };
    } else {
      return {}; // Don't automatically assign a host anymore
    }
  }
  if (config.env === 'development' || process.env.E2E_TEST) {
    if (category === 'opensource') {
      return { CollectiveId: 9805, ParentCollectiveId: 83 }; // Open Source Host Collective
    } else if (category === 'foundation') {
      return { CollectiveId: 9805 };
    } else {
      return {}; // Don't automatically assign a host anymore
    }
  }
  return { id: 1, CollectiveId: 1 };
};

export const isValidEmail = email => {
  if (typeof email !== 'string') {
    return false;
  }
  return email.match(
    /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
  );
};

/**
 * Check if this is an internal email address.
 * Useful for testing emails in localhost or staging
 */
export const isEmailInternal = email => {
  if (!email) {
    return false;
  }
  if (email.match(/(opencollective\.(com|org))$/i)) {
    return true;
  }
  if (email.match(/^xdamman.*@gmail\.com$/)) {
    return true;
  }
  return false;
};

export function capitalize(str) {
  if (!str) {
    return '';
  }
  return str[0].toUpperCase() + str.slice(1).toLowerCase();
}

export function uncapitalize(str) {
  if (!str) {
    return '';
  }
  return str[0].toLowerCase() + str.slice(1);
}

export function pluralize(str, count) {
  if (count <= 1) {
    return str;
  }
  return `${str}s`.replace(/s+$/, 's');
}

export function resizeImage(imageUrl, { width, height, query, defaultImage }) {
  if (!imageUrl) {
    if (defaultImage) {
      imageUrl = defaultImage.substr(0, 1) === '/' ? `${config.host.website}${defaultImage}` : defaultImage;
    } else {
      return null;
    }
  }

  if (imageUrl[0] === '/') {
    imageUrl = `https://opencollective.com${imageUrl}`;
  }

  let queryurl = '';
  if (query) {
    queryurl = `&query=${encodeURIComponent(query)}`;
  } else {
    if (width) {
      queryurl += `&width=${width}`;
    }
    if (height) {
      queryurl += `&height=${height}`;
    }
  }
  return `${config.host.images}/proxy/images/?src=${encodeURIComponent(imageUrl)}${queryurl}`;
}

export function formatArrayToString(arr, conjonction = 'and') {
  if (arr.length === 1) {
    return arr[0];
  }
  if (!arr.slice) {
    return '';
  }
  return `${arr.slice(0, arr.length - 1).join(', ')} ${conjonction} ${arr.slice(-1)}`;
}

export const getDefaultCurrencyPrecision = currency => {
  if (ZERO_DECIMAL_CURRENCIES.includes(currency?.toUpperCase())) {
    return 0;
  } else {
    return 2;
  }
};

export function formatCurrency(amount, currency, precision = 0) {
  amount = amount / 100; // converting cents
  let locale;
  switch (currency) {
    case 'USD':
      locale = 'en-US';
      break;
    case 'EUR':
      locale = 'en-EU';
      break;
    default:
      locale = 'en-US';
  }
  return amount.toLocaleString(locale, {
    style: 'currency',
    currencyDisplay: 'symbol',
    currency,
    minimumFractionDigits: precision,
    maximumFractionDigits: precision,
  });
}

/**
 * @PRE: { USD: 1000, EUR: 6000 }
 * @POST: "€60 and $10"
 */
export function formatCurrencyObject(currencyObj, options = { precision: 0, conjonction: 'and' }) {
  const array = [];
  for (const currency in currencyObj) {
    if (currencyObj[currency] > 0) {
      array.push({
        value: currencyObj[currency],
        str: formatCurrency(currencyObj[currency], currency, options.precision),
      });
    }
  }
  if (array.length === 1) {
    return array[0].str;
  }
  array.sort((a, b) => b.value - a.value);
  return formatArrayToString(
    array.map(r => r.str),
    options.conjonction,
  );
}

export function isUUID(str) {
  return str.length === 36 && str.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);
}

/** Sleeps for MS milliseconds */
export function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export function chunkArray(startArray, chunkSize) {
  let j = -1;
  return startArray.reduce((arr, item, ix) => {
    j += ix % chunkSize === 0 ? 1 : 0;
    arr[j] = [...(arr[j] || []), item];
    return arr;
  }, []);
}

// This generates promises of n-length at a time
// Useful so we don't go over api quota limit on Stripe
export function promiseSeq(arr, predicate, consecutive = 100) {
  return chunkArray(arr, consecutive).reduce((prom, items, ix) => {
    // wait for the previous Promise.all() to resolve
    return prom.then(() => {
      return Promise.all(
        // then we build up the next set of simultaneous promises
        items.map(item => predicate(item, ix)),
      );
    });
  }, Promise.resolve([]));
}

export function parseToBoolean(value) {
  // If value is already a boolean, don't bother converting it
  if (typeof value === 'boolean') {
    return value;
  }

  let lowerValue = value;
  // check whether it's string
  if (lowerValue && (typeof lowerValue === 'string' || lowerValue instanceof String)) {
    lowerValue = lowerValue.trim().toLowerCase();
  }
  if (['on', 'enabled', '1', 'true', 'yes', 1].includes(lowerValue)) {
    return true;
  }
  return false;
}

/**
 * Clean a tags list before inserting to the db. Trim tags, remove empty ones...
 * Will return `null` if the list is empty.
 */
export const cleanTags = tags => {
  if (!tags) {
    return null;
  } else if (typeof tags === 'string') {
    tags = [tags];
  }

  const cleanTagsList = tags
    .filter(t => Boolean(t)) // Remove null values
    .map(t => t.trim()) // Trim tags
    .filter(t => t.length > 0); // Remove empty tags

  return cleanTagsList.length > 0 ? cleanTagsList : null;
};

export const md5 = value => crypto.createHash('md5').update(value).digest('hex');

export const sha512 = value => crypto.createHash('sha512').update(value).digest('hex');

/**
 * Filter `list` with `filterFunc` until `conditionFunc` returns true.
 */
export const filterUntil = (list, filterFunc, conditionFunc) => {
  const result = [];
  for (let i = 0; i < list.length; i++) {
    if (filterFunc(list[i])) {
      result.push(list[i]);
      if (conditionFunc(result)) {
        return result;
      }
    }
  }
  return result;
};

/**
 * @returns boolean: True if `obj` has ony the keys passed in `keys`
 */
export const objHasOnlyKeys = (obj, keys) => {
  const sortedObjKeys = Object.keys(obj).sort();
  const sortedKeys = [...keys].sort();
  return isEqual(sortedObjKeys, sortedKeys);
};

/**
 * Format a datetime object to an ISO date like `YYYY-MM-DD`
 */
export const toIsoDateStr = date => {
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getUTCDate();
  return `${year}-${padStart(month.toString(), 2, '0')}-${padStart(day.toString(), 2, '0')}`;
};

export const getTokenFromRequestHeaders = req => {
  const header = req.headers && req.headers.authorization;
  if (!header) {
    return null;
  }

  const parts = header.split(' ');
  const scheme = parts[0];
  const token = parts[1];
  if (!/^Bearer$/i.test(scheme) || !token) {
    throw new BadRequest('Format is Authorization: Bearer [token]');
  }

  return token;
};

export const sumByWhen = (vector, iteratee, predicate) => sumBy(filter(vector, predicate), iteratee);
