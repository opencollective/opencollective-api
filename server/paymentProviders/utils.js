import slugify from 'limax';

import activities from '../constants/activities';
import { CollectiveType } from '../constants/collectives';
import ExpenseStatus from '../constants/expense-status';
import ExpenseType from '../constants/expense-type';
import { PAYMENT_METHOD_SERVICE, PAYMENT_METHOD_TYPE } from '../constants/paymentMethods';
import { TransactionKind } from '../constants/transaction-kind';
import { TransactionTypes } from '../constants/transactions';
import { getFxRate } from '../lib/currency';
import { crypto } from '../lib/encryption';
import logger from '../lib/logger';
import { toNegative } from '../lib/math';
import { createRefundTransaction } from '../lib/payments';
import models, { Op } from '../models';

export const getVirtualCardForTransaction = async cardId => {
  return models.VirtualCard.findOne({
    where: {
      id: cardId,
    },
    include: [
      { association: 'collective', required: true },
      { association: 'host', required: true },
      { association: 'user' },
    ],
  });
};

export const notifyCollectiveMissingReceipt = async (expense, virtualCard) => {
  expense.collective = expense.collective || (await expense.getCollective());
  expense.fromCollective = expense.fromCollective || (await expense.getFromCollective());
  virtualCard = virtualCard || expense.virtualCard || (await expense.getVirtualCard());

  if (expense.collective.settings?.ignoreExpenseMissingReceiptAlerts === true) {
    return;
  }

  expense.createActivity(
    activities.COLLECTIVE_EXPENSE_MISSING_RECEIPT,
    { id: virtualCard.UserId },
    { ...expense.data, user: virtualCard.user },
  );
};

export const persistTransaction = async (virtualCard, transaction) => {
  // Make sure amount is an absolute value
  const amount = Math.abs(transaction.amount);

  if (amount === 0) {
    return;
  }

  const transactionId = transaction.id;
  const expenseData = { transactionId };
  const UserId = virtualCard.UserId;
  const host = virtualCard.host;
  const collective = virtualCard.collective;
  const vendor = await getOrCreateVendor(transaction.vendorProviderId, transaction.vendorName);
  const currency = transaction.currency || 'USD';
  const hostCurrencyFxRate = await getFxRate(currency, host.currency);
  const description = `Virtual Card charge: ${vendor.name}`;
  const paymentMethod = await host.findOrCreatePaymentMethod(
    PAYMENT_METHOD_SERVICE.STRIPE,
    PAYMENT_METHOD_TYPE.VIRTUAL_CARD,
  );

  // Case when expense is already created after the stripe authorization request event
  // Double check if transaction is a refund because sometimes the refund event is sent before the charge event
  if (transaction.fromAuthorizationId && !transaction.isRefund) {
    const processingExpense = await models.Expense.findOne({
      where: {
        status: ExpenseStatus.PROCESSING,
        VirtualCardId: virtualCard.id,
        data: { authorizationId: transaction.fromAuthorizationId },
      },
    });

    if (processingExpense) {
      // Make sure we update the Expense and ExpenseItem amounts.
      // Sometimes there's a difference between the authorized amount and the charged amount.
      await models.ExpenseItem.update({ amount }, { where: { ExpenseId: processingExpense.id } });
      await processingExpense.update({
        amount,
        paymentMethodId: paymentMethod.id,
        data: { ...expenseData, missingDetails: true, ...transaction.data },
      });
      // Mark Expense as Paid, create activity and don't send notifications
      await processingExpense.markAsPaid({ skipActivity: true });

      await models.Transaction.createDoubleEntry({
        CollectiveId: collective.id,
        FromCollectiveId: vendor.id,
        HostCollectiveId: host.id,
        description,
        type: 'DEBIT',
        currency,
        ExpenseId: processingExpense.id,
        PaymentMethodId: paymentMethod.id,
        amount: toNegative(amount),
        netAmountInCollectiveCurrency: toNegative(amount),
        hostCurrency: host.currency,
        amountInHostCurrency: Math.round(toNegative(amount) * hostCurrencyFxRate),
        paymentProcessorFeeInHostCurrency: 0,
        hostFeeInHostCurrency: 0,
        platformFeeInHostCurrency: 0,
        hostCurrencyFxRate,
        kind: TransactionKind.EXPENSE,
        data: transaction.data,
        clearedAt: transaction.clearedAt,
      });

      const expenseAttachment = await models.ExpenseAttachedFile.findOne({
        where: { ExpenseId: processingExpense.id },
      });

      const expenseItemWithAttachment = await models.ExpenseItem.findOne({
        where: { ExpenseId: processingExpense.id, url: { [Op.ne]: null } },
      });

      if (!expenseAttachment && !expenseItemWithAttachment) {
        await notifyCollectiveMissingReceipt(processingExpense, virtualCard);
      }

      return processingExpense;
    }
  }

  const existingExpense = await models.Expense.findOne({
    where: {
      status: ExpenseStatus.PAID,
      VirtualCardId: virtualCard.id,
      // TODO : only let transactionId in a few months (today : 11/2021) or make a migration to update data on existing expenses and transactions
      data: { [Op.or]: [{ transactionId }, { id: transactionId }, { token: transactionId }] },
    },
  });

  if (existingExpense) {
    logger.warn(`Virtual Card charge already reconciled, ignoring it: ${transactionId}`);
    return;
  }

  if (transaction.isRefund) {
    const existingTransaction = await models.Transaction.findOne({
      where: {
        CollectiveId: collective.id,
        // TODO : only let refundTransactionId in a few months (today : 11/2021) or make a migration to update data on existing expenses and transactions
        data: {
          [Op.or]: [{ refundTransactionId: transactionId }, { id: transactionId }, { token: transactionId }],
        },
      },
    });

    if (existingTransaction) {
      logger.warn(`Virtual Card refund already reconciled, ignoring it: ${transactionId}`);
      return;
    }

    const expense = await models.Expense.findOne({
      where: {
        CollectiveId: collective.id,
        FromCollectiveId: vendor.id,
        VirtualCardId: virtualCard.id,
        type: ExpenseType.CHARGE,
        status: ExpenseStatus.PAID,
        amount,
      },
      order: [['createdAt', 'DESC']],
    });

    if (expense) {
      const [originalCreditTransaction] = await expense.getTransactions({
        where: { type: TransactionTypes.CREDIT, kind: TransactionKind.EXPENSE },
      });
      if (originalCreditTransaction?.amount === amount) {
        if (!originalCreditTransaction.RefundTransactionId) {
          await createRefundTransaction(originalCreditTransaction, 0, { refundTransactionId: transactionId });
        }
        return;
      }
    }

    await models.Transaction.createDoubleEntry({
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      description: `Virtual Card refund: ${vendor.name}`,
      type: 'CREDIT',
      currency,
      amount: amount,
      netAmountInCollectiveCurrency: amount,
      hostCurrency: host.currency,
      amountInHostCurrency: Math.round(amount * hostCurrencyFxRate),
      paymentProcessorFeeInHostCurrency: 0,
      hostFeeInHostCurrency: 0,
      platformFeeInHostCurrency: 0,
      hostCurrencyFxRate,
      isRefund: true,
      kind: TransactionKind.EXPENSE,
      ExpenseId: expense?.id,
      PaymentMethodId: paymentMethod.id,
      data: { refundTransactionId: transactionId },
    });

    return;
  }

  let expense;

  try {
    expense = await models.Expense.create({
      UserId,
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      currency,
      amount,
      description,
      VirtualCardId: virtualCard.id,
      PaymentMethodId: paymentMethod.id,
      lastEditedById: UserId,
      status: ExpenseStatus.PAID,
      type: ExpenseType.CHARGE,
      incurredAt: transaction.incurredAt,
      data: { ...expenseData, missingDetails: true },
    });

    await models.ExpenseItem.create({
      ExpenseId: expense.id,
      incurredAt: transaction.incurredAt,
      CreatedByUserId: UserId,
      amount,
      currency,
    });

    await models.Transaction.createDoubleEntry({
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      description,
      type: 'DEBIT',
      currency,
      ExpenseId: expense.id,
      PaymentMethodId: paymentMethod.id,
      amount: toNegative(amount),
      netAmountInCollectiveCurrency: toNegative(amount),
      hostCurrency: host.currency,
      amountInHostCurrency: Math.round(toNegative(amount) * hostCurrencyFxRate),
      paymentProcessorFeeInHostCurrency: 0,
      hostFeeInHostCurrency: 0,
      platformFeeInHostCurrency: 0,
      hostCurrencyFxRate,
      kind: TransactionKind.EXPENSE,
    });

    await notifyCollectiveMissingReceipt(expense, virtualCard);

    return expense;
  } catch (e) {
    if (expense) {
      await models.Transaction.destroy({ where: { ExpenseId: expense.id } });
      await models.ExpenseItem.destroy({ where: { ExpenseId: expense.id } });
      await expense.destroy();
    }
    throw e;
  }
};

export const getOrCreateVendor = async (vendorProviderId, vendorName) => {
  const slug = vendorProviderId.toString().toLowerCase();
  const hash = crypto.hash(`${slug}${vendorName}`);
  const uniqueSlug = slugify(`${vendorName}-${hash.slice(0, 6)}`);

  const [vendor] = await models.Collective.findOrCreate({
    where: { [Op.or]: [{ slug }, { slug: uniqueSlug }] },
    defaults: { name: vendorName, type: CollectiveType.VENDOR, slug: uniqueSlug },
  });

  if (vendor.name !== vendorName) {
    logger.warn(`Virtual Card: vendor name mismatch for ${vendorProviderId}: '${vendorName}' / '${vendor.name}'`);
  }
  // Update existing vendor to use uniqueSlug.
  if (vendor.slug === slug) {
    await vendor.update({ slug: uniqueSlug });
  }

  return vendor;
};
