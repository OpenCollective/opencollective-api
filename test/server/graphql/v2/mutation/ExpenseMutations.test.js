import { expect } from 'chai';
import { graphqlQueryV2 } from '../../../../utils';
import { fakeCollective, fakeUser } from '../../../../test-helpers/fake-data';
import { randEmail } from '../../../../stores';

const createExpenseMutation = `
mutation createExpense($expense: ExpenseCreate!, $account: AccountInput!) {
  createExpense(expense: $expense, account: $account) {
    id
    legacyId
    invoiceInfo
  }
}`;

describe('server/graphql/v2/mutation/ExpenseMutations', () => {
  describe('createExpense', () => {
    it('creates the expense with the linked attachments', async () => {
      const user = await fakeUser();
      const collective = await fakeCollective();
      const expenseData = {
        description: 'A valid expense',
        type: 'INVOICE',
        invoiceInfo: 'This will be printed on your invoice',
        fromAccount: { legacyId: user.CollectiveId },
        payoutMethod: { type: 'PAYPAL', data: { email: randEmail() } },
        attachments: [{ description: 'A first attachment', amount: 4200 }],
      };

      const result = await graphqlQueryV2(
        createExpenseMutation,
        { expense: expenseData, account: { legacyId: collective.id } },
        user,
      );

      expect(result.errors).to.not.exist;
      expect(result.data).to.exist;
      expect(result.data.createExpense).to.exist;

      const createdExpense = result.data.createExpense;
      expect(createdExpense.invoiceInfo).to.eq(expenseData.invoiceInfo);
    });
  });
});
