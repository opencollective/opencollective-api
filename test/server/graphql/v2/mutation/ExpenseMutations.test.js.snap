// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server/graphql/v2/mutation/ExpenseMutations processExpense PAY Multi-currency expense Pays the expense manually 1`] = `
"
| type   | amount | netAmountInCollectiveCurrency | paymentFee | currency | hostCurrency | hostCurrencyFxRate | To    | From  | Host | isRefund |
| ------ | ------ | ----------------------------- | ---------- | -------- | ------------ | ------------------ | ----- | ----- | ---- | -------- |
| CREDIT | 1700   | 1600                          | -100       | USD      | USD          | 1                  | Payee | Babel | NULL | false    |
| DEBIT  | -1600  | -1700                         | -100       | USD      | USD          | 1                  | Babel | Payee | OSC  | false    |"
`;

exports[`server/graphql/v2/mutation/ExpenseMutations processExpense PAY Multi-currency expense Records a manual payment with an active account 1`] = `
"
| type   | amount | netAmountInCollectiveCurrency | paymentFee | currency | hostCurrency | hostCurrencyFxRate | To    | From  | Host      | isRefund |
| ------ | ------ | ----------------------------- | ---------- | -------- | ------------ | ------------------ | ----- | ----- | --------- | -------- |
| CREDIT | 1700   | 1600                          | -110       | USD      | NZD          | 1.1                | Payee | Babel | PayeeHost | false    |
| DEBIT  | -1600  | -1700                         | -100       | USD      | USD          | 1                  | Babel | Payee | OSC       | false    |"
`;

exports[`server/graphql/v2/mutation/ExpenseMutations processExpense PAY pays 100% of the balance by putting the fees on the payee 1`] = `
"
| type   | kind    | isRefund | To       | From     | amount | amountInHostCurrency | paymentFee | netAmountInCollectiveCurrency |
| ------ | ------- | -------- | -------- | -------- | ------ | -------------------- | ---------- | ----------------------------- |
| DEBIT  | EXPENSE | false    | Webpack  | Facebook | -9425  | -9425                | -575       | -10000                        |
| CREDIT | EXPENSE | false    | Facebook | Webpack  | 10000  | 10000                | -575       | 9425                          |
| DEBIT  | EXPENSE | true     | Facebook | Webpack  | -10000 | -10000               | 575        | -9425                         |
| CREDIT | EXPENSE | true     | Webpack  | Facebook | 9425   | 9425                 | 575        | 10000                         |"
`;

exports[`server/graphql/v2/mutation/ExpenseMutations processExpense PAY pays 100% of the balance by putting the fees on the payee but do not refund processor fees 1`] = `
"
| type   | kind    | isRefund | To       | From     | amount | amountInHostCurrency | paymentFee | netAmountInCollectiveCurrency |
| ------ | ------- | -------- | -------- | -------- | ------ | -------------------- | ---------- | ----------------------------- |
| DEBIT  | EXPENSE | false    | Webpack  | Facebook | -9425  | -9425                | -575       | -10000                        |
| CREDIT | EXPENSE | false    | Facebook | Webpack  | 10000  | 10000                | -575       | 9425                          |
| DEBIT  | EXPENSE | true     | Facebook | Webpack  | -9425  | -9425                | 0          | -9425                         |
| CREDIT | EXPENSE | true     | Webpack  | Facebook | 9425   | 9425                 | 0          | 9425                          |"
`;
